make cli_test wait until it's servers are properly closed

andreimatei opened the pull request:
... so we don't end up with 2 servers running at the same time since
we have some data races between them.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3859)

<!-- Reviewable:end -->

andreimatei committed:
make cli_test wait until it's servers are properly closed

... so we don't end up with 2 servers running at the same time since
we have some data races between them.

bdarnell commented on cli/cli_test.go:
```diff
@@ -35,6 +35,16 @@ type cliTest struct {
+// Stop the server, or wait until stopping is done
+// if it was already initiated by the caller.
+func (c *cliTest) cleanup(wasStopped *bool) {
+	if *wasStopped {
```
I think we can get rid of the `wasStopped` variable. The test will either run to completion or panic, so `cleanup` can be
```
  if p := recover(); p != nil {
   // The test is failing; stop the server and re-panic
   c.Stop()
   panic(p)
  } else {
    // The test ran to completion
    c.WaitForStopped()
  }
```

andreimatei replied:
Well what if the tests would also actually test something and do stuff like `t.FailNow()`?
An alternative I've considered was making Server.Stop() safe to call multiple times, but it didn't look trivial and I wasn't sure I liked it anyway.

bdarnell replied:
There is no `t.FailNow()` in `Example` tests, although that's a concern if there are any non-example CLI tests. You could pass the `t` object (or nil) into cleanup so that you could check `t.Failed()`.

mberhault replied:
why not just `<-c.Stopper().IsStopped()` after the quit command? I kindda like that we're testing it there too. Does calling `c.Stop()` buy us anything?

andreimatei commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[cli/cli_test.go, line 41 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3859#-K814s25Qk4UX7ip8a_Q-r1-41:-K81DGwNGweadVMVzjLx:-169169037) ([raw file](https://github.com/cockroachdb/cockroach/blob/932e1ecf69a4fac46bb899a4aab7bae1fde06b9f/cli/cli_test.go#L41)):*
I've removed the `quit` commands from the examples, and added a specific test to see that it works. Left `WaitForStopped` for that one, everybody else just calls `.Stop`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3859)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[cli/cli_test.go, line 83 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3859#-K81E-BqCY7rg6jmwFYl:-K81E-BqCY7rg6jmwFYm:1943355159) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbc7fcb37d3983b261595b3072243b6230774f1b/cli/cli_test.go#L83)):*
I wouldn't introduce a new method for this since the necessary things are public and it doesn't look likely to be used elsewhere. Just use `<-c.Stopper().IsStopped()`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3859#-:-K81F-UhcvuS-m1q5gst:676834233)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[cli/cli_test.go, line 41 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3859#-K814s25Qk4UX7ip8a_Q-r1-41:-K81HFPmH410FBbv7bor:-123259525) ([raw file](https://github.com/cockroachdb/cockroach/blob/932e1ecf69a4fac46bb899a4aab7bae1fde06b9f/cli/cli_test.go#L41)):*
removing the quit command from the examples was Ben's suggestion, and I agree with it. The output from the quit command seems like it doesn't really belong in the examples?
Plus, not having it allows us to do the shutdown through a defer, so if the control flow in those tests becomes more complex...

---

*[cli/cli_test.go, line 83 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3859#-K81E-BqCY7rg6jmwFYl:-K81Hy3k07ts1lj8062w:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbc7fcb37d3983b261595b3072243b6230774f1b/cli/cli_test.go#L83)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3859)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on cli/cli_test.go:
```diff
@@ -71,8 +77,17 @@ func (c cliTest) RunWithArgs(a []string) {
+	defer leaktest.AfterTest(t)
+	c := newCLITest()
+	c.Run("quit")
+	// Wait until this async command stopps the server.
```
s/pp/p/

petermattis commented:
LGTM

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[cli/cli_test.go, line 41 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3859#-K814s25Qk4UX7ip8a_Q-r1-41:-K84cZxUruLv5c1KFsHC:1958197035) ([raw file](https://github.com/cockroachdb/cockroach/blob/932e1ecf69a4fac46bb899a4aab7bae1fde06b9f/cli/cli_test.go#L41)):*
Not much purpose to the `cleanup` method now. Since `server.TestServer` is embedded, you can just `s/cleanup/Stop/g` below.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3859#-:-K84chFzKXsIvbLW9253:676834233)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[cli/cli_test.go, line 41 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3859#-K814s25Qk4UX7ip8a_Q-r1-41:-K85LHFfKyhXHQdAPPVK:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/932e1ecf69a4fac46bb899a4aab7bae1fde06b9f/cli/cli_test.go#L41)):*
Done.

---

*[cli/cli_test.go, line 84 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3859#-K814s25Qk4UX7ip8a_Q-r3-84:-K85LIJV2hU1lCKETaeP:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/262a8a07c4adf3cf7cce137fb2d48ab0ba9ad5ae/cli/cli_test.go#L84)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3859)*
<!-- Sent from Reviewable.io -->
